pyomo/common/dependencies.py

```diff
@@ -325,20 +392,31 @@ def _finalize_yaml(module, available):
 def _finalize_scipy(module, available):
     if available:
         # Import key subpackages that we will want to assume are present
+        import scipy.stats
         import scipy.sparse
         import scipy.spatial
-        import scipy.stats
 
 def _finalize_pympler(module, available):
     if available:
         # Import key subpackages that we will want to assume are present
         import pympler.muppy
 
+def _finalize_matplotlib(module, available):
+    if available:
+        import matplotlib.pyplot
+
```

Comment:
Since pyplot is imported when finalizing matplotlib, the submodule import is not deferred.  Is that intended?

Reply:
The submodule import is deferred until the main module import is triggered.  So, given:
```python
from pyomo.common.dependencies import matplotlib
plt = matplotlib.pyplot # neither matplotlib nor pyplot are imported
plt.plot([0,1],[2,3]) # both matplotlib and matplotlib.pyplot are imported; plot() works as expected
```
and 
```python
from pyomo.common.dependencies import matplotlib`
plt = matplotlib.pyplot # neither matplotlib nor pyplot are imported
matplotlib.locale # both matplotlib and matplotlib.pyplot are imported
```
The only downside for the `_finalize_matplotlib` approach is that if *any* part of matplotlib is needed, then pyplot will be imported.  When I looked at things, that didn't seem to be a big hit, as pyplot was *always* used if someone was using matplotlib (and this was way easier than implementing a more complex layered deferred import system).